driver-mapping/src/main/java/com/datastax/driver/mapping/Mapper.java

```diff
@@ -104,29 +103,37 @@ Session session() {
         return manager.getSession();
     }
 
-    PreparedStatement getPreparedQuery(QueryType type, Set<ColumnMapper<?>> columns, EnumMap<Option.Type, Option> options) {
-
-        MapperQueryKey pqk = new MapperQueryKey(type, columns, options);
+    ListenableFuture<PreparedStatement> getPreparedQueryAsync(QueryType type, Set<ColumnMapper<?>> columns, EnumMap<Option.Type, Option> options) {
 
+        final MapperQueryKey pqk = new MapperQueryKey(type, columns, options);
         PreparedStatement stmt = preparedQueries.get(pqk);
         if (stmt == null) {
-            synchronized (preparedQueries) {
-                stmt = preparedQueries.get(pqk);
-                if (stmt == null) {
-                    String queryString = type.makePreparedQueryString(tableMetadata, mapper, manager, columns, options.values());
-                    logger.debug("Preparing query {}", queryString);
-                    stmt = session().prepare(queryString);
-                    Map<MapperQueryKey, PreparedStatement> newQueries = new HashMap<MapperQueryKey, PreparedStatement>(preparedQueries);
-                    newQueries.put(pqk, stmt);
-                    preparedQueries = newQueries;
+            String queryString = type.makePreparedQueryString(tableMetadata, mapper, manager, columns, options.values());
+            logger.debug("Preparing query {}", queryString);
+            final SettableFuture<PreparedStatement> future = SettableFuture.create();
+            Futures.addCallback(session().prepareAsync(queryString), new FutureCallback<PreparedStatement>() {
```

Comment:
It's still possible for two threads to prepare the same statement concurrently. It's not that bad, but it prints a warning log.
We can avoid it with a couple of changes:
```java
// store futures instead of plain statements
private final ConcurrentMap<MapperQueryKey, ListenableFuture<PreparedStatement>> preparedQueries
...
SettableFuture<PreparedStatement> future = SettableFuture.create();
ListenableFuture<PreparedStatment> old = preparedQueries.putIfAbsent(pqk, future);
if (old != null) return old;
else {
   // call prepareAsync, complete `future` in the callback
}
```